### DISCUSSION TITLE
Custom layouts and location aware config

### DISCUSSION THREAD

**Discussion by ascheider** [pimcore/pimcore#16176] [Q&A / Support]
Hey folks,

currently i'm updating some pimcore instances from 10 to 11.
Earlier the custom layouts were stored in /var/classes/customlayouts folder.
So the customer or i as developer could modify layout and then just checkin the changes, after deployment the custom layouts were consistent on all environments. It was simple and straightforward solution.
Right now pimcore is using location aware config.
So the layouts are stored in the database in settings_store table. So the layouts needs to be adjusted on each environment or someone needs to export and import them manually via administration.
The second solution is to store the layouts on the filesystem, like here described: https://pimcore.com/docs/platform/Pimcore/Deployment/Configuration_Environments/
But then the layouts are read only in administration and can not be modified by the user anymore.
I also don't see any command or script which can dump the layouts from database to filesystem.
So the question are basically:
1.) How should i process in the project?
2.) Why did pimcore migrate custom layouts to database respectively to file system with read only mode?

it would be really helpful if someone can advice how to process with this in real world project.
Thanks!

**Comment by pradeeppimcore** [comment]
Hi @ascheider 
I am struggling with same issue and trying to figure out the best/intended way to migrate the custom layouts from one env to other env.
Were you able to find out the solution for this ?

Thanks.

**Comment by ascheider** [comment]
@pradeeppimcore yes, it sucks that so important topic is not interesting enough. I'm curious how other people upgrade the instances without having those problems. Maybe only few are using custom layouts.

We implemented event listener which saves custom layout as json on the file system, so we have them in git on the same place.
On deployment we execute the console command which imports those jsons into the database.

Here is the code for the event listener

`class CustomLayoutSubscriber implements EventSubscriberInterface
{

    public static function getSubscribedEvents(): array
    {
        return [
            DataObjectCustomLayoutEvents::PRE_ADD => 'onPreChange',
            DataObjectCustomLayoutEvents::PRE_UPDATE => 'onPreChange',
        ];
    }

    public function onPreChange(CustomLayoutEvent $event): void
    {
        $customLayout = $event->getCustomLayout();
        $json = DataObject\ClassDefinition\Service::generateCustomLayoutJson($customLayout);
        $filename = 'custom_definition_' . $customLayout->getId() . '.json';
        \file_put_contents(PIMCORE_CLASS_DEFINITION_DIRECTORY . '/customlayouts/' .$filename, $json);
    }
}
`
Command code like this:

`$files = glob(PIMCORE_CLASS_DEFINITION_DIRECTORY . '/customlayouts/*.json');
        foreach ($files as $file) {
            if (\preg_match('/.*custom_definition_(.*)\.json$/', $file, $matches)) {
                $id = $matches[1];
                $json = file_get_contents($file);
                $importData = \json_decode($json, true);
                $customLayout = CustomLayout::getById($id);
                if ($customLayout instanceof CustomLayout) {
                    $layout = null;
                    if (!empty($importData['layoutDefinitions'])){
                        $layout = DataObject\ClassDefinition\Service::generateLayoutTreeFromArray($importData['layoutDefinitions'], true);
                    }
                    $customLayout->setLayoutDefinitions($layout);
                    $customLayout->save();
  
                }
            }
        }`

Hope it helps for you.
Maybe write the json file only if not in CLI mode, because $customLayout->save(); will also trigger the save event.

**Reply by natted** [reply]
thanks for posting this @ascheider , I just came across this same issue
